src/source/parserutil.d

```diff
@@ -8,26 +8,32 @@ auto match(Lexer, TokenType)(ref Lexer lexer, TokenType type) {
 		return token;
 	}
 
-	import std.conv, std.string;
-	auto error = token.type == TokenType.Invalid
-		? token.error.toString(lexer.context)
-		: format!"expected '%s', got '%s'."(to!string(type),
-		                                    to!string(token.type));
-
-	import source.exception;
-	throw new CompileException(token.location, error);
+	import std.conv;
+	throw unexpectedTokenError(lexer, to!string(type));
 }
 
 auto unexpectedTokenError(Lexer)(ref Lexer lexer, string expected) {
 	auto token = lexer.front;
 
 	import std.conv, std.string;
-	auto error = token.type == Lexer.TokenType.Invalid
-		? token.error.toString(lexer.context)
-		: format!"expected %s, not `%s`."(expected,
-		                                  token.toString(lexer.context));
+	string error;
+	switch (token.type) {
+		case Lexer.TokenType.Invalid:
+			error = token.error.toString(lexer.context);
+			break;
+		case Lexer.TokenType.End:
+			error = format!"expected %s, not end of input."(expected);
+			break;
+		default:
+			error = format!"expected %s, not `%s`."(expected,
+			                                        token.toString(lexer.context));
+	}
```

Comment:
Because this switch thing just returns some value, if might be worth thinking about making it a function. Not a blocker.